src/platform/platform_linux.cpp

```diff
@@ -154,6 +155,59 @@ void update_bridges(std::map<std::string, mp::NetworkInterfaceInfo>& networks)
 }
 } // namespace
 
+std::pair<QString, QString> mp::platform::Platform::parse_os_release(const QStringList& os_data)
+{
+    const QString id_field = "NAME";
+    const QString version_field = "VERSION_ID";
+
+    QString distro_id = "unknown";
+    QString distro_rel = "unknown";
+
+    for (const QString& line : os_data)
+    {
+        QStringList split = line.split('=', Qt::KeepEmptyParts);
+        if (split.length() == 2 && split[1].length() > 2) // Check for at least 1 char between quotes.
+        {
+            QString val = split[1].mid(1, split[1].length() - 2); // Removing quotes from input string.
+            if (split[0] == id_field)
+                distro_id = val;
+            else if (split[0] == version_field)
+                distro_rel = val;
+        }
+    }
+
+    return std::pair(distro_id, distro_rel);
+}
+
+std::string mp::platform::Platform::read_os_release(const QString& os_release_path)
+{
+    QFile fd(os_release_path);
+    bool exists_os_rel_path = QFileInfo::exists(os_release_path); // Check once, going to disk is expensive.
+    if (!exists_os_rel_path && QFileInfo::exists("/var/lib/snapd/hostfs/etc/os-release"))
+        fd.setFileName("/var/lib/snapd/hostfs/usr/lib/os-release");
+    else if (!exists_os_rel_path && QFileInfo::exists("/var/lib/snapd/hostfs/usr/lib/os-release"))
+        fd.setFileName("/var/lib/snapd/hostfs/etc/os-release");
```

Comment:
I have a couple of concerns about this:
1. The file could exist and still be unreadable. This is not likely, nor is it the situation mentioned in `man 5 os-release`, but we might as well fall back in that case as well. That would amount to going directly to `open` and check whether that succeeds, instead of checking if the file exists.
1. I don't think this function should receive the file to read from and then silently use something else. I understand you want the param for testing purposes, but then this probably needs another layer. How about this:

```cpp
std::unique_ptr<QFile> find_os_release()
{
    const std::array<QString, 3> options{QStringLiteral("/var/lib/snapd/hostfs/etc/os-release"), 
        QStringLiteral("/var/lib/snapd/hostfs/usr/lib/os-release"), 
        QStringLiteral("")};

    auto it = options.cbegin();
    auto ret = std::make_unique<QFile>(*it);
    while (++it != options.cend() && !MP_FILEOPS::open(*ret, QIODevice::ReadOnly | QIODevice::Text))
        ret->setFileName(*it);

    return ret; // need the unique_ptr because QFile is non-copyable
}
```

`MP_FILEOPS::open` replaces `QFile::open` to allow mocking. Then in `read_os_release` you'd receive `const QFile& os_release` and, if open, do you thing, otherwise return the default.

Reply:
Removed `unique_ptr` and switched to `QFile` on stackframe. `unique_ptr` usage necessitates a `dynamic_cast` or release of the `unique_ptr` to create the `QTextStream`. Figuring out tests.

Reply:
Yeah, have a look here: https://github.com/canonical/multipass/blob/1e88479def51630cab9937804f47eee25f820cb0/tests/test_sftpserver.cpp#L1087-L1089